### ISSUE TITLE
Shorter identifiers for better log readability

### ISSUE THREAD

**Issue by pmarini-nc** [strukturag/nextcloud-spreed-signaling#640]
It would be great to have shorter identifiers in the log. This would improve log readability and, according to a simple analysis I did, it wouldn't generate false duplicates if we were to trim the identifiers to 10 letters in a server with around 10-15 active users.

Or am I missing the reason to put the full identifier in the logs?

The analysis I did is the following:
1. output the journald log
journalctl --output=json --output-fields=MESSAGE --unit ncsignaling  > /tmp/ncsignaling.log
2. process it with the following python script:
```
import pandas as pd

def find_largest(ws):
    s_out = ""
    max_len_s = len(s_out)
    for s in ws.split():
        if len(s) >= max_len_s:
            max_len_s = len(s)
            s_out = s
    if max_len_s < 20:
	    s_out = None
    return s_out

df1 = pd.read_json("ncsignaling.log", lines=True)

df1["component"] = df1.MESSAGE.str.split(":").apply(lambda x:x[0])

df2 = df1.query("component=='client.go'")

df2["msg"] = df2.MESSAGE.str.split(":").apply(lambda x:x[2])

df2["identifier"] = df2.msg.apply(find_largest)

df3=df2[df2.identifier.notnull()]

df3["short_id"] = [ identifier[:10] for identifier in df3.identifier.values]

print("Unique Identifiers: %i" % len(df3.identifier.unique()))
print("Unique Short Identifiers: %i" % len(df3.short_id.unique()))
```
The output is:

**Comment by danxuliu** [comment]
Rather than changing the format of the logs an alternative would be to keep the logs just as they are, but use a helper script to process them and shorten the ids as desired. The following script could be used for that (note that it reads the contents from the standard input, so it should be called with something like `cat log-file | python log-shorten-identifiers.py`; it should be possible to pipe `journalctl` too, although I have not verified it):
```
import re
import sys

SESSION_ID_LENGTH = 10

sessions = dict()

for log in sys.stdin:
    log = log.strip()

    if ' Register ' in log:
        [session, privateSession] = re.findall('Register .* from .* in .* \(.*\) (.*) \(private=(.*)\)', log)[0]

        sessions[session] = session[:SESSION_ID_LENGTH]
        sessions[privateSession] = privateSession[:SESSION_ID_LENGTH]

    # Use RTT messages to also add the session in case the register message
    # is previous to the beginning of the log.
    if ' has RTT of ' in log:
        [session] = re.findall('Client (.*) has RTT of ', log)
        
        sessions[session] = session[:SESSION_ID_LENGTH]

    for session in sessions.keys():
        log = log.replace(session, sessions[session])
    
    if ' Unregister ' in log:
        [session, privateSession] = re.findall('Unregister (.*) \(private=(.*)\)', log)[0]

        if session in sessions:
            sessions.pop(session)
        if privateSession in sessions:
            sessions.pop(privateSession)

    print(log)
```

**Comment by pmarini-nc** [comment]
Thanks for providing this script @danxuliu!

This could be a nice workaround, but with its problems from a maintainability point of view. When operating multiple signaling servers, it must be uploaded and updated in all of them. Even if the maintenance is low, it shouldn't be taken for granted that the script works as expected with new releases of the signaling server.

**Comment by danxuliu** [comment]
> This could be a nice workaround, but with its problems from a maintainability point of view. When operating multiple signaling servers, it must be uploaded and updated in all of them. Even if the maintenance is low, it shouldn't be taken for granted that the script works as expected with new releases of the signaling server.

Yes, I totally agree that the script has its drawbacks. It was just provided as a work around in case the IDs are kept with its current length :-)

> On one side this would be the first and only service that I've seen so far among many (database, web server, cache ..) that would need an external "string shortener" script, on the other I didn't see any feedback from other signaling server administrators since this request has been created.

Personally when analysing logs I just search for the first ~10 characters of the session id and assume that the rest will match 😅 But at least for me the long ids have never been a problem 🤷 (although it is true that if I copy some relevant messages to discuss them with other people in a chat I usually shorten the IDs, but I usually just keep the original ones if I am analysing them myself).

> What is the reason to have these very long session strings?

My guess would be that it was done to play safe and "ensure" that there would be no clashes between sessions even on large clusters of signaling servers with thousands or maybe even millions of users. But maybe it is just the default length of the algorithm used to encode the sessions and that is it, without any further reason. I am afraid that I do not know, @fancycode would be the best person to answer it :-)

**Comment by fancycode** [comment]
Session ids contain encoded information about the session of a user. This includes a hash to prevent modifications and they are encrypted to prevent extraction of information. This is the reason they are rather long (which already improved a bit with #850).

Codec:
https://github.com/strukturag/nextcloud-spreed-signaling/blob/master/sessionid_codec.go

Contained data:
https://github.com/strukturag/nextcloud-spreed-signaling/blob/53f736cb2ec21cdb3f6ea805b8ebe5af0160aefe/session.proto#L30

In the past, the idea was that this allows different instances in a cluster to decrypt them and read the contents even if the session was created on a different instance (using the same de-/encryption key). However as the clustering is done through GRPC now, this might no longer be necessary and we could just generate completely random strings.

I will think about it and leave this open as a reminder.
